libcudacxx/include/cuda/std/__type_traits/apply_cv.h

```diff
@@ -20,63 +20,176 @@
 #  pragma system_header
 #endif // no system header
 
-#include <cuda/std/__type_traits/is_const.h>
-#include <cuda/std/__type_traits/is_volatile.h>
-#include <cuda/std/__type_traits/remove_reference.h>
-#include <cuda/std/cstddef>
-
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-template <class _Tp,
-          class _Up,
-          bool = is_const<__libcpp_remove_reference_t<_Tp>>::value,
-          bool = is_volatile<__libcpp_remove_reference_t<_Tp>>::value>
-struct __apply_cv
+struct __apply_cv_
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = _Tp;
+};
+
+struct __apply_cv_c
 {
-  typedef _LIBCUDACXX_NODEBUG_TYPE _Up type;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const _Tp;
 };
 
-template <class _Tp, class _Up>
-struct __apply_cv<_Tp, _Up, true, false>
+struct __apply_cv_v
 {
-  typedef _LIBCUDACXX_NODEBUG_TYPE const _Up type;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = volatile _Tp;
 };
 
-template <class _Tp, class _Up>
-struct __apply_cv<_Tp, _Up, false, true>
+struct __apply_cv_cv
 {
-  typedef volatile _Up type;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const volatile _Tp;
 };
 
-template <class _Tp, class _Up>
-struct __apply_cv<_Tp, _Up, true, true>
+struct __apply_cv_lr
 {
-  typedef const volatile _Up type;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = _Tp&;
 };
 
-template <class _Tp, class _Up>
-struct __apply_cv<_Tp&, _Up, false, false>
+struct __apply_cv_clr
 {
-  typedef _Up& type;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const _Tp&;
 };
 
-template <class _Tp, class _Up>
-struct __apply_cv<_Tp&, _Up, true, false>
+struct __apply_cv_vlr
 {
-  typedef const _Up& type;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = volatile _Tp&;
 };
 
-template <class _Tp, class _Up>
-struct __apply_cv<_Tp&, _Up, false, true>
+struct __apply_cv_cvlr
 {
-  typedef volatile _Up& type;
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const volatile _Tp&;
 };
 
-template <class _Tp, class _Up>
-struct __apply_cv<_Tp&, _Up, true, true>
+struct __apply_cv_rr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = _Tp&&;
+};
+
+struct __apply_cv_crr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const _Tp&&;
+};
+
+struct __apply_cv_vrr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = volatile _Tp&&;
+};
+
+struct __apply_cv_cvrr
+{
+  template <class _Tp>
+  using __call _LIBCUDACXX_NODEBUG_TYPE = const volatile _Tp&&;
+};
+
+#ifndef _CCCL_NO_VARIABLE_TEMPLATES
+template <class>
+extern __apply_cv_ __apply_cvr;
```

Comment:
This adds a ton of indirection, but I do not really see where we safe any compile time 

Can you elaborate in what regard this performs better than the previous approach

Reply:
this is essentially the implementation from `__tuple_dir/apply_cv.h` moved to `__type_traits/apply_cv.h`, with a minor tweak. instead of instantiating templates `is_lvalue_reference`, `is_const`, and `is_volatile` (and adding `is_rvalue_reference`), it instantiates just one: `__apply_cvr`.